### PR TITLE
Snapshot status

### PR DESCRIPTION
**Pull request by AndreKurait**
### Description
Implements several changes for performance and usability:
* Implements Snapshot Status
* Adds Snapshot Max Throughput parameter to Create
* Makes Snapshot Create Asynchronous with `--wait` option
* Modifies RFS to support ES 7.17
* Changes the RFS ECS Worker to use the right document migration main class
* Adds Bash-Completion to migration console with support for console-link library
* Fixes RFS logging pattern in Cloudwatch
_______
* Category: New Features
* Why these changes are required? Usability improvements for RFS and Migration Console
* What is the old behavior before changes and new behavior after changes?

### Issues Resolved
[MIGRATIONS-1792](https://opensearch.atlassian.net/browse/MIGRATIONS-1792)

Is this a backport? If so, please add backport PR # and/or commits #

### Testing
Lots of manual testing locally and in the cloud

Follow up for more documentation and E2E cloud testing https://opensearch.atlassian.net/browse/MIGRATIONS-1808

### Check List
- [x] New functionality includes testing
  - [x] All tests pass, including unit test, integration test and doctest
- [ ] New functionality has been documented
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: RFS/src/main/java/com/rfs/common/SnapshotShardUnpacker.java [hunk 1]
```diff
@@ -72,40 +70,10 @@ public Path unpack() {
         }
     }
 
-    @Override
-    public void close() {
```
Comment (mikaylathompson) [inline RFS/src/main/java/com/rfs/common/SnapshotShardUnpacker.java:76]: do we care about these things not getting cleaned up?

File: TrafficCapture/dockerSolution/src/main/docker/migrationConsole/lib/console_link/console_link/models/snapshot.py [hunk 2]
```diff
@@ -38,55 +41,209 @@ def status(self, *args, **kwargs) -> CommandResult:
...
+
+class ClusterSnapshotState(Enum):
+    SUCCESS = "SUCCESS"
+    FAILED = "FAILED"
+    PARTIAL = "PARTIAL"
+    STARTED = "STARTED"
+    IN_PROGRESS = "IN_PROGRESS"
+
+
+def convert_snapshot_state_to_status(snapshot_state: str) -> Tuple[SnapshotStatus, str]:
```
Comment (mikaylathompson) [inline TrafficCapture/dockerSolution/src/main/docker/migrationConsole/lib/console_link/console_link/models/snapshot.py:121]: I'm not sure I understand why we should have State & Status and what the difference is. Should we just use these States throughout?

File: TrafficCapture/dockerSolution/src/main/docker/migrationConsole/loadServicesFromParameterStore.sh [hunk 3]
```diff
@@ -29,3 +29,14 @@ if [ $? -ne 0 ]; then
...
+# Generate bash completion script
+console completion bash > /usr/share/bash-completion/completions/console
+
+# Source the completion script to enable it for the current session
+source /usr/share/bash-completion/completions/console
+
+# Add sourcing of the completion script to .bashrc for persistence across sessions
+echo '. /etc/bash_completion' >> ~/.bashrc
+
+echo "Bash completion for console command has been set up and enabled."
```
Comment (mikaylathompson) [inline TrafficCapture/dockerSolution/src/main/docker/migrationConsole/loadServicesFromParameterStore.sh:42]: Thanks for adding this!

**Review by mikaylathompson — APPROVED** [review]
Looks good! A couple questions, but no blockers.

File: TrafficCapture/dockerSolution/src/main/docker/migrationConsole/lib/console_link/console_link/models/snapshot.py (hunk 2 above)
Comment (AndreKurait) [reply TrafficCapture/dockerSolution/src/main/docker/migrationConsole/lib/console_link/console_link/models/snapshot.py:193]: State is the cluster response enum whereas Status is what we define. I had originally defined it when we were also storing status in a separate doc.

This helps us maintain a consistent Status despite state renaming across ES/OS versions

File: RFS/src/main/java/com/rfs/common/SnapshotShardUnpacker.java (hunk 1 above)
Comment (AndreKurait) [reply RFS/src/main/java/com/rfs/common/SnapshotShardUnpacker.java:76]: No, we want to recycle tasks between shards with ECS. When this was in there, we were deleting the shard date before we reindexed it

File: TrafficCapture/dockerSolution/src/main/docker/migrationConsole/lib/console_link/console_link/models/snapshot.py (hunk 2 above)
Comment (peternied) [reply TrafficCapture/dockerSolution/src/main/docker/migrationConsole/lib/console_link/console_link/models/snapshot.py:181]: This doesn't feel right, we are embedding version handling deep in the console.  I understand getting this in position for the demo, but this blends the responsibilities of the system quite a bit - it is going to make our codebase hard to maintain very quickly.

@AndreKurait What do you think of this, how should we capture the responsibility questions?

File: TrafficCapture/dockerSolution/src/main/docker/migrationConsole/lib/console_link/console_link/models/snapshot.py
Comment (peternied) [inline TrafficCapture/dockerSolution/src/main/docker/migrationConsole/lib/console_link/console_link/models/snapshot.py:1]: Lets add test cases for the new scenarios, get that coverage number up a bit more

**Review by peternied — CHANGES_REQUESTED** [review]

File: TrafficCapture/dockerSolution/src/main/docker/migrationConsole/lib/console_link/console_link/models/snapshot.py
Comment (AndreKurait) [reply TrafficCapture/dockerSolution/src/main/docker/migrationConsole/lib/console_link/console_link/models/snapshot.py:1]: I'll move this to use the string status

Comment (AndreKurait) [reply TrafficCapture/dockerSolution/src/main/docker/migrationConsole/lib/console_link/console_link/models/snapshot.py:1]: Added Tests

File: TrafficCapture/dockerSolution/src/main/docker/migrationConsole/lib/console_link/console_link/models/snapshot.py (hunk 2 above)
Comment (AndreKurait) [reply TrafficCapture/dockerSolution/src/main/docker/migrationConsole/lib/console_link/console_link/models/snapshot.py]: Moved this to just return the state string that the cluster responds with

**Review by peternied — APPROVED** [review]
Thanks for the updates @AndreKurait

**Pull request merged**